### ISSUE TITLE
[governance] Gas optimization in state function of Governor contract

### ISSUE THREAD

**Issue by TheSnakeWitcher** [OpenZeppelin/openzeppelin-contracts#4201]
Maybe there is something that I am not aware of, but I think that
the line

```solidity
ProposalCore storage proposal = _proposals[proposalId]
```

can be substituted by

```solidity
ProposalCore memory proposal = _proposals[proposalId]
```

Rationale is that `proposal` is never mutated, and we are reading
unnecessarily from storage here

```solidity
        if (proposal.executed) {
            return ProposalState.Executed;
        }

        if (proposal.canceled) {
            return ProposalState.Canceled;
        }
```

Full code next

```solidity
    /**
     * @dev See {IGovernor-state}.
     */
    function state(uint256 proposalId) public view virtual override returns (ProposalState) {
        ProposalCore storage proposal = _proposals[proposalId];

        if (proposal.executed) {
            return ProposalState.Executed;
        }

        if (proposal.canceled) {
            return ProposalState.Canceled;
        }

        uint256 snapshot = proposalSnapshot(proposalId);

        if (snapshot == 0) {
            revert("Governor: unknown proposal id");
        }

        uint256 currentTimepoint = clock();

        if (snapshot >= currentTimepoint) {
            return ProposalState.Pending;
        }

        uint256 deadline = proposalDeadline(proposalId);

        if (deadline >= currentTimepoint) {
            return ProposalState.Active;
        }

        if (_quorumReached(proposalId) && _voteSucceeded(proposalId)) {
            return ProposalState.Succeeded;
        } else {
            return ProposalState.Defeated;
        }
    }
```

**Comment by Amxx** [comment]
Hello

The rational for using storage instead of memory is that we don't want to load all the slots unless it is necessary.

A proposal currently uses 3 slots:
- one for `voteStart` and `proposer`
- one for `voteEnd`
- one for `executed` and `cancelled`.

Currently, we indeed have to sload the last one twice. Once for checking execute and once for checking canceled.
But if we used memory as your propose we would read all 3 slots. Sure we would avoid one "hot" sload (which would save 100 gas) ... but we would also be causing 2 unecessary sloads. That is (200 ~ 4000) more gas depending on the situation.

We are planning to improve packing of the proposal details in 5.0, but until then there is no real room for optimisations.
